java/java-common/src/main/java/org/cloudburstmc/protocol/java/JavaClient.java

```diff
@@ -0,0 +1,174 @@
+import java.net.InetSocketAddress;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Consumer;
+
+@Getter
+public class JavaClient extends Java {
+    private InetSocketAddress remoteAddress;
+    private JavaClientSession session;
+    private int timeout;
+    private boolean closed = false;
+    private boolean connected = false;
+
+    @Setter private JavaEventHandler<JavaClientSession> handler;
+
+    private Bootstrap bootstrap;
+
+    public JavaClient() {
+        this(null, EventLoops.commonGroup());
+    }
+
+    public JavaClient(InetSocketAddress bindAddress) {
+        this(bindAddress, EventLoops.commonGroup());
+    }
+
+    public JavaClient(EventLoopGroup eventLoopGroup) {
+        this(null, eventLoopGroup);
+    }
+
+    public JavaClient(InetSocketAddress bindAddress, EventLoopGroup eventLoopGroup) {
+        this(bindAddress, eventLoopGroup, 30_000);
+    }
+
+    public JavaClient(EventLoopGroup eventLoopGroup, int timeout) {
+        this(null, eventLoopGroup, timeout);
+    }
+
+    public JavaClient(@Nullable InetSocketAddress bindAddress, EventLoopGroup eventLoopGroup, int timeout) {
+        super(bindAddress, eventLoopGroup);
+
+        this.timeout = timeout;
+    }
+
+    @Override
+    public InetSocketAddress getBindAddress() {
+        if (this.bindAddress == null) {
+            return (InetSocketAddress) this.session.getChannel().localAddress();
+        }
+        return super.getBindAddress();
+    }
+
+    @Override
+    protected void onTick() {
+        if (this.session.isConnected()) {
+            this.session.tick();
+        }
+    }
+
+    @Override
+    public CompletableFuture<Void> bind() {
+        Preconditions.checkNotNull(this.eventLoopGroup, "Event loop group was null");
+        this.bootstrap = new Bootstrap()
+                .channel(EventLoops.getChannelType().getSocketChannel())
+                .group(this.eventLoopGroup)
+                .localAddress(this.bindAddress)
+                .option(ChannelOption.TCP_NODELAY, true)
+                .option(ChannelOption.CONNECT_TIMEOUT_MILLIS, this.timeout)
+                .handler(new ChannelInitializer<Channel>() {
+                    @Override
+                    public void initChannel(Channel channel) throws Exception {
+                        JavaClientSession session = new JavaClientSession(JavaClient.this, (InetSocketAddress) channel.localAddress(), channel, channel.eventLoop());
+                        channel.pipeline()
+                                .addLast("timeout", new ReadTimeoutHandler(JavaClient.this.timeout))
+                                .addLast("splitter", new VarInt21FrameDecoder())
+                                .addLast("prepender", new VarInt21LengthFieldPrepender())
+                                .addLast("encoder", new PacketEncoder(session))
+                                .addLast("decoder", new PacketDecoder(session, JavaPacketType.Direction.CLIENTBOUND))
```

Comment:
TODO: Add HAProxy support.

Reply:
It doesn't necessarily have to be part of this library. As long as we expose the pipeline, it will enable users to add whatever codecs they want.